enterprise/coderd/workspaceagents_test.go

```diff
@@ -73,6 +82,103 @@ func TestBlockNonBrowser(t *testing.T) {
 	})
 }
 
+func TestReinitializeAgent(t *testing.T) {
+	t.Parallel()
+
+	tempAgentLog := testutil.CreateTemp(t, "", "testReinitializeAgent")
+
+	if !dbtestutil.WillUsePostgres() {
+		t.Skip("dbmem cannot currently claim a workspace")
+	}
+
+	// GIVEN a live enterprise API with the prebuilds feature enabled
+	client, db, user := coderdenttest.NewWithDatabase(t, &coderdenttest.Options{
+		Options: &coderdtest.Options{
+			DeploymentValues: coderdtest.DeploymentValues(t, func(dv *codersdk.DeploymentValues) {
+				dv.Experiments.Append(string(codersdk.ExperimentWorkspacePrebuilds))
+			}),
+		},
+		LicenseOptions: &coderdenttest.LicenseOptions{
+			Features: license.Features{
+				codersdk.FeatureWorkspacePrebuilds: 1,
+			},
+		},
+	})
+
+	// GIVEN a template, template version, preset and a prebuilt workspace that uses them all
+	presetID := uuid.New()
+	tv := dbfake.TemplateVersion(t, db).Seed(database.TemplateVersion{
+		OrganizationID: user.OrganizationID,
+		CreatedBy:      user.UserID,
+	}).Preset(database.TemplateVersionPreset{
+		ID: presetID,
+	}).Do()
+
+	r := dbfake.WorkspaceBuild(t, db, database.WorkspaceTable{
+		OwnerID:    prebuilds.SystemUserID,
+		TemplateID: tv.Template.ID,
+	}).Seed(database.WorkspaceBuild{
+		TemplateVersionID: tv.TemplateVersion.ID,
+		TemplateVersionPresetID: uuid.NullUUID{
+			UUID:  presetID,
+			Valid: true,
+		},
+	}).WithAgent(func(a []*proto.Agent) []*proto.Agent {
+		a[0].Scripts = []*proto.Script{
+			{
+				DisplayName: "Prebuild Test Script",
+				Script:      fmt.Sprintf("sleep 5; printenv | grep 'CODER_AGENT_TOKEN' >> %s; echo '---\n' >> %s", tempAgentLog.Name(), tempAgentLog.Name()), // Make reinitialization take long enough to assert that it happened
+				RunOnStart:  true,
+			},
+		}
+		return a
+	}).Do()
+
+	// GIVEN a running agent
+	logDir := t.TempDir()
+	inv, _ := clitest.New(t,
+		"agent",
+		"--auth", "token",
+		"--agent-token", r.AgentToken,
+		"--agent-url", client.URL.String(),
+		"--log-dir", logDir,
+	)
+	clitest.Start(t, inv)
+
+	// GIVEN the agent is in a happy steady state
+	waiter := coderdtest.NewWorkspaceAgentWaiter(t, client, r.Workspace.ID)
+	waiter.WaitFor(coderdtest.AgentsReady)
+
+	// WHEN a workspace is created that can benefit from prebuilds
+	ctx := testutil.Context(t, testutil.WaitShort)
+	workspace, err := client.CreateUserWorkspace(ctx, user.UserID.String(), codersdk.CreateWorkspaceRequest{
+		TemplateVersionID:       tv.TemplateVersion.ID,
+		TemplateVersionPresetID: presetID,
+		Name:                    "claimed-workspace",
+	})
+	require.NoError(t, err)
+
+	db.UpdateProvisionerJobWithCompleteByID(ctx, database.UpdateProvisionerJobWithCompleteByIDParams{
+		ID:        workspace.LatestBuild.ID,
+		UpdatedAt: time.Now(),
+		CompletedAt: sql.NullTime{
+			Valid: true,
+			Time:  time.Now(),
+		},
+	})
```

Comment:
Either you are running a provisioner daemon, in which case it should be responsible for completing the build job, or you are not, in which case, you need something else to send the pubsub kick that the build is done.

Reply:
You have a provisioner daemon running, which should publish this already on CompleteJob